### ISSUE TITLE
Unexpected attention results using phoneme based cleaners

### ISSUE THREAD

**Issue by SanjaESC** [mozilla/TTS#534]
Branch: Master ([540d811dd5](https://github.com/mozilla/TTS/tree/540d811dd52b5598a7cd21cbbcf197b0bfbeab62))

Hi, 

trying to train a multi-speaker model using the current master branch with phoneme computation I encounter the following weird attention alignment problem. 

![Bildschirmfoto von 2020-09-30 21-17-37](https://user-images.githubusercontent.com/6319070/94729733-69b3c700-0362-11eb-8bcb-fa0da3a63e2c.png)
![Bildschirmfoto von 2020-09-30 21-28-54](https://user-images.githubusercontent.com/6319070/94730749-fb700400-0363-11eb-8cae-010d5c711826.png)

I am using the basic_german_cleaner for text cleaning. I also tried the https://github.com/repodiac/german_transliterate from @repodiac but it yielded the same results.
Corresponding config [config.zip](https://github.com/mozilla/TTS/files/5307795/config.zip)

Switching back to character based training resolves the problem. Any ideas?

**Comment by george-roussos** [comment]
I just had a look at the german cleaner and it looks okay -- what dataset are you using? If you try e.g. `echo "lächeln" | phonemize -l de` on the terminal, do you get the correct output?

I would also try a run with restoring the German single speaker TTS. It should help with alignments

**Comment by thorstenMueller** [comment]
We used german cleaner by @repodiac during training „thorsten“ dataset and didn’t encounter these weird alignment issues. We use the german cleaner when running tts-server and have no issues on pronouncing german umlauts.

I’ve generated a test sentence and uploaded it on soundcloud.

https://soundcloud.com/thorsten-mueller-395984278/tts-thorsten-de-testsatz-mit-deutschen-umlauten

```
Dieser Testsatz dient zur Überprüfung der Aussprache der deutschen Umlaute mit deutschem Bereinigungsskript, wie Ä in Änderungswunsch, Ü in Überraschungsei, Ö in Öffentlichkeit oder auch Straße mit ß sollte gehen. 
```

**Comment by repodiac** [comment]
Hi @SanjaESC, if you can nail it down to a wrong phoneme encoding based on cleaning I am happy to help (and fix) out :) As @thorstenMueller  pointed out, we do not know those issues yet - if you find a wrong transliteration I can also only strongly encourage you to file an issue on my github page: https://github.com/repodiac/german_transliterate/issues

**Comment by SanjaESC** [comment]
@george-roussos Using `echo "Die süße Hündin läuft in die Höhle des Bären, der sie zum Kaffeekränzchen eingeladen hat." | phonemize -l de` yields the following result `diː zyːsə hʏndɪn lɔøft ɪn diː høːlə dɛs bɛːrən dɛɾ ziː tsʊm kafeːkɾɛntsçən aɪnɡəlɑːdən hat` which seems fine to me. I also printed out the phoneme conversion while it was caching the initial run and they all seemed fine.

@thorstenMueller @repodiac Pretty sure your cleaner isn't the root cause of the problem, since I get the same results with the default cleaner. I'll try some more experiments and report back.

PS: The character based multi-speaker model is training just fine.

**Comment by thorstenMueller** [comment]
@SanjaESC my output of phonemize -l de is near, but little bit different than your output:

thorsten:
diː zyːsə hʏndɪn lɔøft iːn diː høːlə deːs bɛːrən deːɾ ziː tsʊm kafeːkɾɛntsçən aɪnɡəlɑːdən hɑːt

sanja:
diː zyːsə hʏndɪn lɔøft ɪn diː høːlə dɛs bɛːrən dɛɾ ziː tsʊm kafeːkɾɛntsçən aɪnɡəlɑːdən hat

**Comment by SanjaESC** [comment]
@thorstenMueller Interesting! Any idea why that could be? Did you make any changes to espeak? 
My phonemizer version is 2.2.1. 

**Update:**
The attention aligns using a single speaker.
![Bildschirmfoto von 2020-10-01 12-25-19](https://user-images.githubusercontent.com/6319070/94798630-2bf68300-03e2-11eb-8e6c-8ea3751ce81e.png)

**Comment by erogol** [comment]
you might be using GMM attention. Can you post your config.json?

**Comment by SanjaESC** [comment]
> you might be using GMM attention. Can you post your config.json?



    // AUDIO PARAMETERS
    "audio":{
        // stft parameters
        "fft_size": 1024,         // number of stft frequency levels. Size of the linear spectogram frame.
        "win_length": 1024,      // stft window length in ms.
        "hop_length": 256,       // stft window hop-lengh in ms.
        "frame_length_ms": null, // stft window length in ms.If null, 'win_length' is used.
        "frame_shift_ms": null,  // stft window hop-lengh in ms. If null, 'hop_length' is used.

        // Audio processing parameters
        "sample_rate": 22050,   // DATASET-RELATED: wav sample-rate.
        "preemphasis": 0.0,     // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.
        "ref_level_db": 20,     // reference level db, theoretically 20db is the sound of air.

        // Silence trimming
        "do_trim_silence": true,// enable trimming of slience of audio as you load it. LJspeech (true), TWEB (false), Nancy (true)
        "trim_db": 60,          // threshold for timming silence. Set this according to your dataset.

        // Griffin-Lim
        "power": 1.5,           // value to sharpen wav signals after GL algorithm.
        "griffin_lim_iters": 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.

        // MelSpectrogram parameters
        "num_mels": 80,         // size of the mel spec frame.
        "mel_fmin": 0.0,        // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!
        "mel_fmax": 8000.0,     // maximum freq level for mel-spec. Tune for dataset!!
        "spec_gain": 20.0,

        // Normalization parameters
        "signal_norm": true,    // normalize spec values. Mean-Var normalization if 'stats_path' is defined otherwise range normalization defined by the other params.
        "min_level_db": -100,   // lower bound for normalization
        "symmetric_norm": true, // move normalization to range [-1, 1]
        "max_norm": 4.0,        // scale normalization to range [-max_norm, max_norm] or [0, max_norm]
        "clip_norm": true,      // clip normalized values into the range.
        "stats_path": null    // DO NOT USE WITH MULTI_SPEAKER MODEL. scaler stats file computed by 'compute_statistics.py'. If it is defined, mean-std based notmalization is used and other normalization params are ignored
    },

    // VOCABULARY PARAMETERS
    // if custom character set is not defined,
    // default set in symbols.py is used
    "characters":{
        "pad": "_",
        "eos": "~",
        "bos": "^",
        "characters": "ABCDEFGHIJKLMNOPQRSTUVWXYZÄÖÜabcdefghijklmnopqrstuvwxyzäöüß!'(),-.:;? ",
        "punctuations":"!'(),-.:;? ",
        "phonemes":"iyɨʉɯuɪʏʊeøɘəɵɤoɛœɜɞʌɔæɐaɶɑɒᵻʘɓǀɗǃʄǂɠǁʛpbtdʈɖcɟkɡqɢʔɴŋɲɳnɱmʙrʀⱱɾɽɸβfvθðszʃʒʂʐçʝxɣχʁħʕhɦɬɮʋɹɻjɰlɭʎʟˈˌːˑʍwɥʜʢʡɕʑɺɧɚ˞ɫ"
    },

    // DISTRIBUTED TRAINING
    "distributed":{
        "backend": "nccl",
        "url": "tcp:\/\/localhost:54321"
    },

    "reinit_layers": [],    // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.

    // TRAINING
    "batch_size": 32,       // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.
    "eval_batch_size":16,
    "r": 7,                 // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled.
    "gradual_training": [[0, 7, 64], [1, 5, 64], [50000, 3, 32], [130000, 2, 32], [290000, 2, 32]], //set gradual training steps [first_step, r, batch_size]. If it is null, gradual training is disabled. For Tacotron, you might need to reduce the 'batch_size' as you proceeed.
    "loss_masking": true,         // enable / disable loss masking against the sequence padding.
    "ga_alpha": 10.0,        // weight for guided attention loss. If > 0, guided attention is enabled.
    "apex_amp_level": null, // level of optimization with NVIDIA's apex feature for automatic mixed FP16/FP32 precision (AMP), NOTE: currently only O1 is supported, and use "O1" to activate.

    // VALIDATION
    "run_eval": true,
    "test_delay_epochs": 400,  //Until attention is aligned, testing only wastes computation time.
    "test_sentences_file": "/home/akorolev/master/projects/TTS/sentences.txt",  // set a file to load sentences to be used for testing. If it is null then we use default english sentences.

    // OPTIMIZER
    "noam_schedule": false,        // use noam warmup and lr schedule.
    "grad_clip": 1.0,              // upper limit for gradients for clipping.
    "epochs": 1000,                // total number of epochs to train.
    "lr": 0.0001,                  // Initial learning rate. If Noam decay is active, maximum learning rate.
    "wd": 0.000001,                // Weight decay weight.
    "warmup_steps": 4000,          // Noam decay steps to increase the learning rate from 0 to "lr"
    "seq_len_norm": false,         // Normalize eash sample loss with its length to alleviate imbalanced datasets. Use it if your dataset is small or has skewed distribution of sequence lengths.

    // TACOTRON PRENET
    "memory_size": -1,             // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.
    "prenet_type": "bn",           // "original" or "bn".
    "prenet_dropout": false,       // enable/disable dropout at prenet.

    // TACOTRON ATTENTION
    "attention_type": "original",  // 'original' or 'graves'
    "attention_heads": 4,          // number of attention heads (only for 'graves')
    "attention_norm": "sigmoid",   // softmax or sigmoid.
    "windowing": false,            // Enables attention windowing. Used only in eval mode.
    "use_forward_attn": false,     // if it uses forward attention. In general, it aligns faster.
    "forward_attn_mask": false,    // Additional masking forcing monotonicity only in eval mode.
    "transition_agent": false,     // enable/disable transition agent of forward attention.
    "location_attn": true,         // enable_disable location sensitive attention. It is enabled for TACOTRON by default.
    "bidirectional_decoder": false,  // use https://arxiv.org/abs/1907.09006. Use it, if attention does not work well with your dataset.
    "double_decoder_consistency": true,  // use DDC explained here https://erogol.com/solving-attention-problems-of-tts-models-with-double-decoder-consistency-draft/
    "ddc_r": 7,                           // reduction rate for coarse decoder.

    // STOPNET
    "stopnet": true,               // Train stopnet predicting the end of synthesis.
    "separate_stopnet": true,      // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.

    // TENSORBOARD and LOGGING
    "print_step": 25,       // Number of steps to log training on console.
    "tb_plot_step": 100,    // Number of steps to plot TB training figures.
    "print_eval": false,     // If True, it prints intermediate loss values in evalulation.
    "save_step": 10000,      // Number of training steps expected to save traninpg stats and checkpoints.
    "checkpoint": true,     // If true, it saves checkpoints per "save_step"
    "tb_model_param_stats": false,     // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.

    // DATA LOADING
    "text_cleaner": "basic_german_cleaners",
    "enable_eos_bos_chars": false, // enable/disable beginning of sentence and end of sentence chars.
    "num_loader_workers": 2,        // number of training data loader processes. Don't set it too big. 4-8 are good values.
    "num_val_loader_workers": 2,    // number of evaluation data loader processes.
    "batch_group_size": 0,  //Number of batches to shuffle after bucketing.
    "min_seq_len": 6,       // DATASET-RELATED: minimum text length to use in training
    "max_seq_len": 150,     // DATASET-RELATED: maximum text length

    // PATHS
    "output_path": "/home/akorolev/master/projects/TTS/Trainings/",

    // PHONEMES
    "phoneme_cache_path": "/home/akorolev/master/projects/TTS/phoneme_cache/",  // phoneme computation is slow, therefore, it caches results in the given folder.
    "use_phonemes": true,           // use phonemes instead of raw characters. It is suggested for better pronounciation.
    "phoneme_language": "de",     // depending on your target language, pick one from  https://github.com/bootphon/phonemizer#languages

    // MULTI-SPEAKER and GST
    "use_speaker_embedding": true,      // use speaker embedding to enable multi-speaker learning.
    "use_external_speaker_embedding_file": false, // if true, forces the model to use external embedding per sample instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "external_speaker_embedding_file": null, // if not null and use_external_speaker_embedding_file is true, it is used to load a specific embedding file and thus uses these embeddings instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "use_gst": true,       			    // use global style tokens
    "gst":	{			                // gst parameter if gst is enabled
        "gst_style_input": null,        // Condition the style input either on a
                                        // -> wave file [path to wave] or
                                        // -> dictionary using the style tokens {'token1': 'value', 'token2': 'value'} example {"0": 0.15, "1": 0.15, "5": -0.15}
                                        // with the dictionary being len(dict) <= len(gst_style_tokens).
        "gst_embedding_dim": 512,
        "gst_num_heads": 4,
        "gst_style_tokens": 10
	},

    // DATASETS
    "datasets":   // List of datasets. They all merged and they get different speaker_ids.
        [
            {
                "name": "gothic_tts",
                "path":  "/home/akorolev/master/projects/data/SpeechData/tts/DialogSpeech/",
                "meta_file_train": null,
                "meta_file_val": null
            }
        ]

**Comment by erogol** [comment]
Try changing ```prenet_type:original```. And switch it to bn after some training.

**Comment by SanjaESC** [comment]
> Try changing `prenet_type:original`. And switch it to bn after some training.

Sorry I maybe should have stated that i initially tried training it with `prenet_type: original` and had the same results. 
What confuses me is that training on graphemes works just fine. There the attention aligns already at around ~3k steps.
Using a single-speaker + phonemes works.

I will test some more and report back if I find something. Could be obv. also a problem on my end. ^^

**Comment by george-roussos** [comment]
Try a run restoring Thorsten's TTS. It may help 👍

**Comment by SanjaESC** [comment]
Update:

Using the current [dev](https://github.com/mozilla/TTS/tree/48a40c47304ddf8670b539174497cd0acab0f0e5) branch works. Still not sure what the root of the caused problem is.
